Adds Swedish Mac ISO and ANSI keymaps

cfarvidson opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

The Swedish keycodes didn't work for me on my Mac so I created ones that do work.

I created https://github.com/qmk/qmk_firmware/pull/12937 but was asked to create complete sets of ISO and ANSI files.

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Adds working swedish keycodes for mac

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

cfarvidson commented:
Any news on this @drashna @fauxpark ? 😊

fauxpark commented on quantum/keymap_extras/keymap_swedish_osx_ansi.h:
```diff
@@ -0,0 +1,233 @@
+
+/*
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬─────┐
+ * │ < │ 1 │ 2 │ 3 │ 4 │ 5 │ 6 │ 7 │ 8 │ 9 │ 0 │ + │ ´ │     │
```
Sublime Text 4 now renders NBSPs... differently:
![image](https://user-images.githubusercontent.com/4781841/122659368-db66c600-d1ba-11eb-8f11-6360cc005474.png)
So my clever technique to make these readable even with whitespace shown, is not so clever anymore :(

If you could change them all to spaces? (I'll fix the rest of the headers at some point later...)

cfarvidson replied:
👍 

I have updated the files and replaced the NBSPs with spaces.

fauxpark commented on quantum/keymap_extras/keymap_swedish_osx_ansi.h:
```diff
@@ -0,0 +1,233 @@
+
+/* Alted symbols
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬─────┐
+ * │ ≤ │ © │ @ │ £ │ $ │ ∞ │   │ | │ [ │ ] │ ≈ │ ± │   │     │
```
Alt+1 and Alt+\ on the ANSI variant appear to be swapped:
![image](https://user-images.githubusercontent.com/4781841/123678286-79c4ec80-d889-11eb-8ec2-e26c357ac188.png)

cfarvidson replied:
> Alt+1 and Alt+\ on the ANSI variant appear to be swapped:

I have fixed the missing §. But I don't understand the swapped part. On the mac layout the @ is on Alt + 1.

fauxpark commented on quantum/keymap_extras/keymap_swedish_osx_ansi.h:
```diff
@@ -0,0 +1,233 @@
+
+/* Shift+Alted symbols
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬─────┐
+ * │ ≥ │ ¡ │   │ ¥ │ ¢ │ ‰ │   │ \ │ { │ } │ ≠ │ ¿ │   │     │
```
Looks like there are some missing symbols here? or are they duplicates?
![image](https://user-images.githubusercontent.com/4781841/123678711-0374ba00-d88a-11eb-8f4d-e466678c291f.png)

cfarvidson replied:
I have pushed a couple of new commits that adds missing keys and I've checked for duplicates. 🙂

cfarvidson replied:
Yes. The missing ones should all be duplicates

cfarvidson commented:
Thank you so much for looking at this @fauxpark ! 🙂 

I have to admit that I have been struggling with this. I haven't found the best process for testing both the ansi and the iso versions. The way I do it is that I have my own keymap in another branch that I rebase on this to see that it compiles and the do some manual tests.

So any tips would be appreciated.

That said I think that I've added the missing keys, double checked for duplicates in both files and corrected an error.

And I think that I should be able to merge now.

fauxpark commented on quantum/keymap_extras/keymap_swedish_osx_ansi.h:
```diff
@@ -0,0 +1,241 @@
+#define SE_DCIR S(A(SE_I))    // ˆ
+#define SE_NARP S(A(SE_P))    // ∏
+#define SE_RNGA S(A(SE_ARNG)) // ˚
+                              // ^ (duplicate)
```
```suggestion
```

fauxpark commented on quantum/keymap_extras/keymap_swedish_osx_ansi.h:
```diff
@@ -0,0 +1,241 @@
+#define SE_DLQU S(A(SE_COMM)) // „
+#define SE_MDDT S(A(SE_DOT))  // ·
+#define SE_MDSH S(A(SE_MINS)) // —
+
```
```suggestion
```

cfarvidson replied:
Fixed. 👍

fauxpark reviewed (changes requested):
Finally got around to sitting down and really looking at this.
These suggestions match what I see in the Input Sources tab when I plug in an external keyboard to my MBP, select the Swedish layout, and hold Shift/Option/Shift+Option, with the keyboard set as ANSI and ISO from the "Change Keyboard Type" button.

cfarvidson commented on quantum/keymap_extras/keymap_swedish_osx_iso.h:
```diff
@@ -0,0 +1,240 @@
+// Row 1 
+#define SE_PILC A(SE_SECT) // ¶
+#define SE_COPY A(SE_1)    // ©
+#define SE_AT   A(SE_2)    // @
```
Maybe I'm missunderstanding something but I think that this is correct.

This is a picture of my swedish macbook keyboard and you send an @ with `A(SE_2)` ![IMG_2496](https://user-images.githubusercontent.com/7823236/124229350-c0c31400-db0d-11eb-8d7a-361ec19445f7.jpeg)

fauxpark replied:
That is the `Swedish - Pro` layout, which I assume is a special arrangement only for Macs (or perhaps the Macbook Pro specifically, given the name). According to [this SE question](https://apple.stackexchange.com/questions/60831/whats-the-difference-between-the-swedish-and-swedish-pro-keyboard-layouts), the only difference is apparently the swapped @ and ™.

cfarvidson replied:
I would like to keep the the files as they are so that they match the `Swedish - Pro` layout. 

I've never used anything else, and I really believe that it comes as the default setting on swedish macs. Otherwise the mac keyboards doesn't really work with the labels on the keys.

And if @ and ™ must be swapped the keycodes will not match the actual keys printed, and that was the origin story behind this PR. 🙂

fauxpark replied:
That's fine, as long as you do so in a separate set of headers eg. `keymap_swedish_pro_osx_[ansi|iso].h`.

fauxpark replied:
It's also worth noting that these headers are kinda useless if you're typing on a Mac keyboard - they don't run QMK. Which is why I'm not terribly concerned that they don't match up.

cfarvidson replied:
> That's fine, as long as you do so in a separate set of headers eg. keymap_swedish_pro_osx_[ansi|iso].h.

Great I'll fix more files.

> It's also worth noting that these headers are kinda useless if you're typing on a Mac keyboard - they don't run QMK. Which is why I'm not terribly concerned that they don't match up.

I don't understand how you mean. The `Swedish - Pro` layout affects QMK keyboards? Wouldn't I have to specifically set the QMK keyboard to the legacy `Swedish` layout to get `@` working correctly otherwise?

fauxpark replied:
The defines in these headers, on their own, don't really do a whole lot. They're just aliases to the "US ANSI" keycodes as defined in the HID spec. If you add `SE_ODIA` to your keymap and press it, QMK isn't sending an `ö`, it sends the same value as it would for `KC_SCLN`, and only if your OS is set up to use the Swedish keyboard layout will you get an `ö`. So you can think of them more as keycaps; they just help with readability in your `keymap.c`.

QMK obviously doesn't run on builtin Macbook keyboards, or other Apple-branded keyboards. Since this `Swedish - Pro` layout appears to be Macbook Pro-specific, it warrants an additional set of files if someone wants to use it on an external keyboard. Otherwise, my assumption was that users would simply pick the normal Swedish layout from the Input Sources list.

cfarvidson commented on quantum/keymap_extras/keymap_swedish_osx_iso.h:
```diff
@@ -0,0 +1,240 @@
+#define SE_UNDS S(SE_MINS) // _
+
+/* Alted symbols
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───────┐
+ * │ ¶ │ © │ @ │ £ │ $ │ ∞ │   │ | │ [ │ ] │ ≈ │ ± │   │       │
+ * ├───┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─────┤
+ * │     │ • │ Ω │ é │ ® │ † │ µ │ ü │ ı │ œ │ π │ ˙ │ ~ │     │
+ * ├─────┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┐    │
+ * │      │  │ ß │ ∂ │ ƒ │ ¸ │ ˛ │ √ │ ª │ ﬁ │ ø │ æ │ ™ │    │
+ * ├────┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴───┴────┤
+ * │    │ ≤ │ ÷ │   │ ç │ ‹ │ › │ ‘ │ ’ │ ‚ │ … │ – │          │
+ * ├────┼───┴┬──┴─┬─┴───┴───┴───┴───┴───┴──┬┴───┼───┴┬────┬────┤
+ * │    │    │    │                        │    │    │    │    │
+ * └────┴────┴────┴────────────────────────┴────┴────┴────┴────┘
```
I'm fixing the whitespace and the uppercase. See my comment bellow for the switching of `SE_AT`.

cfarvidson commented on quantum/keymap_extras/keymap_swedish_osx_ansi.h:
```diff
@@ -0,0 +1,241 @@
+
+/* Shift+Alted symbols
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬─────┐
+ * │ ≥ │ ¡ │ ” │ ¥ │ ¢ │ ‰ │   │ \ │ { │ } │ ≠ │ ¿ │   │     │
```
Good catch on the duplicate `”` and missing `¶`!

Thanks!

cfarvidson commented on quantum/keymap_extras/keymap_swedish_osx_iso.h:
```diff
@@ -0,0 +1,240 @@
+#define SE_FI   A(SE_L)    // ﬁ
+#define SE_OSTR A(SE_ODIA) // ø
+#define SE_AE   A(SE_ADIA) // æ
+#define SE_TM   A(SE_QUOT) // ™
```
See my other comment about `@`

cfarvidson commented on quantum/keymap_extras/keymap_swedish_osx_ansi.h:
```diff
@@ -0,0 +1,241 @@
+
+/* Alted symbols
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬─────┐
+ * │ ≤ │ © │ @ │ £ │ $ │ ∞ │ § │ | │ [ │ ] │ ≈ │ ± │   │     │
```
See my other comment about `@`

cfarvidson commented on quantum/keymap_extras/keymap_swedish_osx_ansi.h:
```diff
@@ -0,0 +1,241 @@
+// Row 1 
+#define SE_LTEQ A(SE_LABK) // ≤
+#define SE_COPY A(SE_1)    // ©
+#define SE_AT   A(SE_2)    // @
```
See my other comment about `@`

cfarvidson commented:
Thank you @fauxpark for looking at this!

Very appreciated to have another set of eyes on this. You found some good stuff!

I have gone through them all. The only thing I left out was the one about swapping `@` and `™`. I really can't get my head around that one because on my physical keyboard it makes sense. (But I may have been staring att this for a few to many hours 😉  😂 )

Thanks again for all the help!

cfarvidson commented on quantum/keymap_extras/keymap_swedish_osx_ansi.h:
```diff
@@ -0,0 +1,241 @@
+#define SE_YEN  S(A(SE_3))    // ¥
+#define SE_CENT S(A(SE_4))    // ¢
+#define SE_PERM S(A(SE_5))    // ‰
+#define SE_BSLS S(A(SE_7))    // (backslash)
```
This should be resolved.

cfarvidson commented:
@fauxpark I've created 2 new files for the `Swedish Pro` layout and added the changes to the `Swedish` layout.

cfarvidson commented:
@fauxpark Any chance of getting this merged? 🙂

fauxpark commented on quantum/keymap_extras/keymap_swedish_osx_ansi.h:
```diff
@@ -0,0 +1,240 @@
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬─────┐
+ * │ ≤ │ © │ ™ │ £ │ $ │ ∞ │ § │ | │ [ │ ] │ ≈ │ ± │   │     │
+ * ├───┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬───┤
+ * │     │ • │ Ω │ É │ ® │ † │ µ │ Ü │ ı │ Œ │ π │ ˙ │ ~ │ ™ │
```
```suggestion
 * │     │ • │ Ω │ É │ ® │ † │ µ │ Ü │ ı │ Œ │ π │ ˙ │ ~ │ @ │
```

fauxpark commented on quantum/keymap_extras/keymap_swedish_pro_osx_ansi.h:
```diff
@@ -0,0 +1,240 @@
+ */
+// Row 1
+#define SE_RABK S(SE_LABK) // >
+#define SE_EXLM S(SE_1)    // ! 
```
```suggestion
#define SE_EXLM S(SE_1)    // !
```

fauxpark commented on quantum/keymap_extras/keymap_swedish_osx_iso.h:
```diff
@@ -0,0 +1,239 @@
+ * ├───┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬───┤
+ * │     │ • │ Ω │ É │ ® │ † │ µ │ Ü │ ı │ Œ │ π │ ˙ │ ~ │   │
+ * ├─────┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┐  │
+ * │      │  │ ß │ ∂ │ ƒ │ ¸ │ ˛ │ √ │ ª │ ﬁ │ Ø │ Æ │ ™ │  │
```
```suggestion
 * │      │  │ ß │ ∂ │ ƒ │ ¸ │ ˛ │ √ │ ª │ ﬁ │ Ø │ Æ │ @ │  │
```

fauxpark commented on quantum/keymap_extras/keymap_swedish_osx_iso.h:
```diff
@@ -0,0 +1,239 @@
+
+/* Alted symbols
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬─────┐
+ * │ ¶ │ © │ @ │ £ │ $ │ ∞ │   │ | │ [ │ ] │ ≈ │ ± │   │     │
```
```suggestion
 * │ ¶ │ © │ ™ │ £ │ $ │ ∞ │   │ | │ [ │ ] │ ≈ │ ± │   │     │
```

fauxpark reviewed (changes requested):
Just a couple small tweaks to the layout comments, otherwise LGTM

cfarvidson commented:
@fauxpark 

> Just a couple small tweaks to the layout comments, otherwise LGTM

Done. 🙂

cfarvidson commented:
@fauxpark I'm I supposed to do something more with this? I'm not really sure on the next step in getting this merged.

🙂

drashna commented:
> @fauxpark I'm I supposed to do something more with this? I'm not really sure on the next step in getting this merged.
> 
> 🙂

Mostly for one of us to merge it.  

Also, thanks!
